### ISSUE TITLE
ExperimentDataPipe data shuffling

### ISSUE THREAD

**Issue by atolopko-czi** [chanzuckerberg/cellxgene-census#538]
Shuffling of training data load via `ExperimentDataPipe` can currently be accomplished using PyTorch's [Shuffler](https://pytorch.org/data/beta/generated/torchdata.datapipes.iter.Shuffler.html#torchdata.datapipes.iter.Shuffler) pipe. However, this operation can only shuffle data within a fixed batch size (since data is streamed). If the shuffle batch size is large enough, the randomization may be acceptable. But this depends upon the ordering of the Census data (along the obs axis) and upon the model architecture being trained. E.g., model is being trained on multiple tissues, but the data from a given dataset is for a single tissue and larger than the shuffle batch size, then Shuffler will not produce a randomized order where different tissues are interleaved.

So it may be beneficial to support an initial full shuffling along the obs axis of the query result. This would allow Shuffler's batch size to be less of a concern, since Shuffler's batch-size shuffling would occur on an already-randomized ordering of Census data.

This could be implemented by ExperimentDataLoader fairly easily, since the obs soma_joinids are retrieved fully before X data iteration commences. The training-time retrieval of X/obs data would occur in the global randomized order.

**Comment by atolopko-czi** [comment]
With local census on EBS volume, c5.9xlarge instance (36 VCPU, 72GB RAM) to iterate through all human, primary data:

| shuffling | soma_chunk_size | elapsed | max mem (GiB) |
|--------|--------|--------|--|
| no            | 100,000                  |  | |
| yes          | 100,000                  | 2:07:47 | 10.363 |
| no          | 1,000,000                  | 0:27:23 | 13.744 |
| yes          | 1,000,000                  | 0:30:12 | 25.419 |
